### ISSUE TITLE
How to guarantee the relaxation from {0, 1} to [0, 1] results in optimized s contains values close to 0 or 1, but not 0.5?

### ISSUE THREAD

**Issue by Tonniia** [KaidiXu/GCN_ADV_Train#8]
I wonder if the relaxation from $s\in\{0,1\}$ to $s\in[0,1]$ can guarantee the optimized $s$ contains values close to 0 or 1. If most of the values are like $0.45, 0.55$, is it reasonable to use the sampling algorithm 1 to recover it back to binary variable?
thanks in advance.

**Comment by KaidiXu** [comment]
No, we cannot guarantee $s$ to be close to 0 or 1. But the values themselves can be interpreted as probabilities, so it's still reasonable to sample based on them whatever the distribution.

**Comment by Sutongtong233** [comment]
Thank you for your quick response. I have another question. Is relaxation from {0, 1} to [0, 1] and sampling have some reference from combinatorial optimization theoretical support? I do not find citations around Eq.(8)...(Even though this is an intuitive strategy)

**Comment by Tonniia** [comment]
Thanks a lot! I have another question about space complexity. Since the modified_adj does not sparse during optimization, if I want to use this attack on a large graph, storing such a large dense adjacency matrix will be difficult. Is there any strategy for dealing with this problem?

**Comment by KaidiXu** [comment]
You are right, this is a limitation.
